plugins/modules/source_control/gitlab/gitlab_integration.py

```diff
@@ -0,0 +1,546 @@
+    elements: str
+  state:
+    description:
+      - Create or delete an integration.
+      - Possible values are C(present) and C(absent).
+    default: "present"
+    choices: ["present", "absent"]
+    type: str
+'''
+
+EXAMPLES = '''
+# Setup email on push for this project
+- name: Email me on push
+  community.general.gitlab_integration:
+    api_url: https://gitlab.com
+    api_token: foobar
+    project: 123456
+    integration: emails-on-push
+    params:
+      recipients: foo@example.com
+      disable_diffs: true
+    state: present
+  delegate_to: localhost
+
+# This will always be set to change because a non-null token is mandatory
+- name: Trigger packagist update on push events (only)
+  community.general.gitlab_integration:
+    api_url: https://gitlab.com
+    api_token: foobar
+    project: foo/proj
+    integration: packagist
+    events: ["push"]
+    params:
+      username: foo
+      token: bar
+      server: https://packagist.org
+  no_log: True
+  delegate_to: localhost
+
+- Idempotency is only partially provided since GitLab does
+  not expose secret parameters like tokens or passwords.
+  See U(https://gitlab.com/gitlab-org/gitlab-ce/issues/46313)
+'''
+
+RETURN = '''
+---
+integration:
+  description: A dict containing key/value pairs representing GitLab integration
+  returned: success
+  type: dict
+  sample:
+    id: 40812345
+    push_events: true
+    issues_events: true
+    confidential_issues_events: true
+    merge_requests_events: true
+    tag_push_events: true
+    note_events: true
+    confidential_note_events: true
+    job_events: true
+    pipeline_events: true
+    wiki_page_events: true
+    recipients: me@example.com
+    disable_diffs: null
+    send_from_committer_email: null
+    title: Emails on push
+    created_at: '2018-05-13T03:08:07.943Z'
+    updated_at: '2018-05-13T03:09:17.651Z'
+    active: true
+    properties:
+      send_from_committer_email: null
+      disable_diffs: null
+      recipients: me@example.com
+    project_id: 1234567
+state:
+  description: A string indicating whether the integration was "created" or "changed"
+  returned: success
+  type: str
+  sample: created
+'''
+
+import traceback
+
+from ansible.module_utils.six import iteritems
+
+GITLAB_IMP_ERR = None
+try:
+    import gitlab
+    HAS_GITLAB_PACKAGE = True
+except ImportError:
+    GITLAB_IMP_ERR = traceback.format_exc()
+    HAS_GITLAB_PACKAGE = False
+
+from ansible.module_utils.api import basic_auth_argument_spec
+from ansible.module_utils.basic import AnsibleModule, missing_required_lib
+from ansible.module_utils.common.text.converters import to_native
+
+from ansible_collections.community.general.plugins.module_utils.gitlab import auth_argument_spec, gitlab_authentication
+
+'''
+This dict has been auto-generated 2022/02/17 from https://gitlab.com/gitlab-org/gitlab-foss/-/blob/master/lib/api/helpers/integrations_helpers.rb
+following these steps:
+1. Download above-mentioned file.
+2. Prepend:
```

Comment:
question: this is cool, but it adds a lot of custom code that needs to be manually executed periodically. You should be able to get most of the details with python-gitlab using `_service_attrs` "for free" from here: https://github.com/python-gitlab/python-gitlab/blob/bd1ecdd5ad654b01b34e7a7a96821cc280b3ca67/gitlab/v4/objects/services.py#L34

Does it make sense to just reuse that? Obviously that list needs to be updated too, but so would this module's eventually - maybe best to keep it in one place? Feel free to open a PR in the library and ping me.

You could then dynamically add `no_log` if you just handle `no_log_keys` in the python code here as you already have the constant. You would lose the `type` constraint for sub-elements, unless that's also added upstream in the library.

Reply:
In the long run, that would clearly be better if this `_service_attrs` were provided by the library as a kind of harcoded API documentation. But even then, this Ansible module wouldn't be allowed to rely on it because it needs to work with older versions of python-gitlab which do not provide it.

Overall, the presence of this object is aimed at validating properly the module's arguments (instead of letting the user providing an opaque unvalidated `dict`), but the whole thing depends on the final decision of @felixfontein